graphql-rewrite/src/entry_point.rs

```diff
@@ -0,0 +1,274 @@
+                });
+                args.push(PyToken {
+                    kind: P::Dollar,
+                    value: "$".into(),
+                    position: None,
+                });
+                args.push(PyToken {
+                    kind: P::Name,
+                    value: varname.into(),
+                    position: None,
+                });
+                args.push(PyToken {
+                    kind: P::Colon,
+                    value: ":".into(),
+                    position: None,
+                });
+                args.push(PyToken {
+                    kind: P::Name,
+                    value: "String".into(),
+                    position: None,
+                });
+                args.push(PyToken {
+                    kind: P::Bang,
+                    value: "!".into(),
+                    position: None,
+                });
+                continue;
+            }
+            _ => {}
+        }
+        tmp.push(PyToken::new((token, pos))?);
+    }
+    insert_args(&mut tokens, &oper.insert_variables, args);
+    tokens.extend(tmp);
+
+    for tok in src_tokens.drain(src_tokens.len()) {
+        tokens.push(PyToken::new(tok)?);
+    }
+
+    return Ok(Entry {
+        key: join_tokens(&tokens),
+        variables,
+        defaults: BTreeMap::new(),
+        tokens,
+        end_pos,
+    })
+}
+
+fn join_tokens<'a, I: IntoIterator<Item=&'a PyToken>>(tokens: I) -> String {
+    let mut buf = String::new();
+    let mut needs_whitespace = false;
+    for token in tokens {
+        match (token.kind, needs_whitespace) {
+            // space before puncutators is optional
+            (PyTokenKind::ParenL, true) => {},
+            (PyTokenKind::ParenR, true) => {},
+            (PyTokenKind::Spread, true) => {},
+            (PyTokenKind::Colon, true) => {},
+            (PyTokenKind::Equals, true) => {},
+            (PyTokenKind::At, true) => {},
+            (PyTokenKind::BracketL, true) => {},
+            (PyTokenKind::BracketR, true) => {},
+            (PyTokenKind::BraceL, true) => {},
+            (PyTokenKind::BraceR, true) => {},
+            (PyTokenKind::Pipe, true) => {},
+            (PyTokenKind::Bang, true) => {},
+            (_, true) => buf.push(' '),
+            (_, false) => {},
+        }
+        buf.push_str(&token.value);
+        needs_whitespace = match token.kind {
+            PyTokenKind::Dollar => false,
+            PyTokenKind::Bang => false,
+            PyTokenKind::ParenL => false,
+            PyTokenKind::ParenR => false,
+            PyTokenKind::Spread => false,
+            PyTokenKind::Colon => false,
+            PyTokenKind::Equals => false,
+            PyTokenKind::At => false,
+            PyTokenKind::BracketL => false,
+            PyTokenKind::BracketR => false,
+            PyTokenKind::BraceL => false,
+            PyTokenKind::BraceR => false,
+            PyTokenKind::Pipe => false,
+            PyTokenKind::Int => true,
```

Comment:
Wouldn't it be better to just match only the ones that need `true` / `unreachable()` and let the rest fall through to `false`?

Reply:
Fallthrough is bad, because if you add another token, you must carefully consider if it needs true or false here. And compiler would not help you find places where there is a fall through branch.

Reply:
Understood.